File: java/leetcode/GraphBipartite.java
//Leetcode problem 785 Is Graph Bipartitie
//Solution written by Xuqiang Fang on 21 May, 2018
import java.util.HashMap;
import java.util.Map;
import java.util.LinkedList;
import java.util.Queue;

class Solution{
    public boolean isBipartite(int[][] graph) {
        Map<Integer,Integer> visited = new HashMap<>();
        Queue<Integer> queue = new LinkedList<>();
        for(int i=0; i<graph.length; ++i){
            if(!visited.containsKey(i)){
                visited.put(i, 0);
                queue.offer(i);
                while(!queue.isEmpty()){
                    //the size of each level
                    int size = queue.size();
                    while(size-- > 0){
                        int x = queue.poll();
                        int[] p = graph[x];
                        for(int j : p){
                            if(!visited.containsKey(j)){
                                //for this level, all nodes should belong to the other group
                                visited.put(j, 1-visited.get(x));
                                queue.offer(j);
                            }
                            //and if there is a contradiction, it is not bipartite
                            else if(visited.get(j) == visited.get(x)){
                                return false;
                            }
                        }
                    }
                }
            }
        }
        return true;
    }
    
}

public class GraphBipartite{
	public static void main(String[] args){
		Solution s = new Solution();
        int[][] graph = {{1,3},{0,2},{1,3},{0,2}};
        int[][] graph2= {{1,2,3},{0,2},{0,1,3},{0,2}};
        int[][] graph3 = {
        {},{2,4,6},{1,4,8,9},{7,8},{1,2,8,9},{6,9},{1,5,7,8,9},{3,6,9},{2,3,4,6,9},{2,4,5,6,7,8}
        };
        System.out.println(s.isBipartite(graph));
        System.out.println(s.isBipartite(graph2));
        System.out.println(s.isBipartite(graph3));
	}
}

File: py/leetcode/GraphBipartite.py
'''
Leetcode problem No 785 Is Graph Bipartite
Solution written by Xuqiang Fang on 21 May, 2018
'''

class Solution(object):
    def isBipartite(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: bool
        """
        dic = {}
        queue = []
        for i in range(len(graph)):
            if not i in dic:
                dic[i] = 0
                queue.append(i)
                while len(queue) > 0:
                    size = len(queue)
                    while size > 0:
                        x = queue.pop(0)
                        tmp = graph[x]
                        for y in tmp:
                            if not y in dic:
                                dic[y] = 1-dic[x]
                                queue.append(y)
                            elif dic[y] == dic[x]:
                                return False
                        size -= 1
        return True

def main():
    s = Solution()
    graph = [[1,3], [0,2], [1,3], [0,2]]
    graph2 = [[1,2,3], [0,2], [0,1,3], [0,2]]
    graph3 = [[],[2,4,6],[1,4,8,9],[7,8],[1,2,8,9],[6,9],[1,5,7,8,9],[3,6,9],[2,3,4,6,9],[2,4,5,6,7,8]]
    print(s.isBipartite(graph))
    print(s.isBipartite(graph2))
    print(s.isBipartite(graph3))

main()
